### PR TITLE
Account for member injection in assisted types

### LINKED ISSUE #342: AssistedInject-annotated classes's factories don't account for member injection
Given this example

```kotlin
class Example @AssistedInject constructor() {
  @Inject lateinit var injected: String
}
```

The generated factory is missing a member injection call

```kotlin
public class Example_Factory(
  // private val param0: Provider<String> // <-- Missing parameter input
) : Factory<Example> {
  public override fun get(): Example {
    val instance = newInstance()
    // Example_MembersInjector.injectInjected(instance, param0.get()) // <-- Missing injection
    return instance
  }
}
```

The relevant `MembersInjector` is correctly generated, just not accounted for here.

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #342 and also leaves some important consolidated logic toe-holds along the way for helping with #341, #340, #339, and #343

### CODE REVIEW

**Comment by ZacSweers** [comment]
I found it was trivial to fix property setters while I was at this, so resolved #340 with 340. Can move to another PR though if you really want!

**Comment by ZacSweers** [comment]
Tested this locally and confirmed this resolves issues in our project and works properly with Dagger's generated components trying to access them 👍

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.KtClassBody
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.psiUtil.visibilityModifierTypeOrDefault
+
+// TODO
+//  Include methods: https://github.com/square/anvil/issues/339
+//  Include superclass: https://github.com/square/anvil/issues/343
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt:20]: Superclass won't work with Psi APIs.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+internal fun KtClassOrObject.injectedMembers(module: ModuleDescriptor) = children
+  .asSequence()
+  .filterIsInstance<KtClassBody>()
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt:23]: Does this work with nested classes?

File: compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 3]
```diff
@@ -180,6 +180,13 @@ public fun Any.invokeGet(vararg args: Any?): Any {
   return method.invoke(this, *args)
 }
 
+@ExperimentalAnvilApi
+public fun Any.getPropertyValue(name: String): Any {
+  val propField = this::class.java.getDeclaredField(name)
+    .apply { isAccessible = true }
+  return propField.get(this)
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt:187]: oh that's neat

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt [hunk 4]
```diff
@@ -220,6 +231,12 @@ public final class InjectClass_MembersInjector implements MembersInjector<Inject
         .invoke(null, injectInstanceStatic, Pair(Pair("e", 1), setOf("f")))
       membersInjector.staticInjectMethod("set")
         .invoke(null, injectInstanceStatic, setOf { _: List<String> -> listOf("g") })
+      // From dagger, it will just see a plain old "set" method rather than a field
+      val name = if (useDagger) "setSetterAnnotated" else "setterAnnotated"
```
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt:235]: it's not a problem in practice, but for the sake of looking it up in this test we need to support both. Dagger-compiler only sees the bean setter name

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt (hunk 1 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt:20]: Are you sure? We do some superclass searching in our internal plugin 🤔

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt (hunk 2 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt:23]: I can add a test. The implementation was the same in all three places so if it worked in one it shouldn't be different here

File: compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt]: cc66813 and also make it work for fields in 387e8d7

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt (hunk 2 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt:23]: Test in 31c01cc

**Comment by ZacSweers** [comment]
Both failures look like CI flakes

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (vRallev) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt:259]: Why is this not an issue in practice? Dagger generates this method and other code could reference the static methods. Dagger generates:
```java
  public static void injectSetSetterAnnotated(InjectClass instance, Map<String, String> p0) {
    instance.setSetterAnnotated(p0);
  }
```
Anvil generates:
```kotlin
    @JvmStatic
    @InjectedFieldSignature("com.squareup.test.InjectClass.setterAnnotated")
    public fun injectSetterAnnotated(instance: InjectClass, setterAnnotated: @JvmSuppressWildcards
        Map<String, String>): Unit {
      instance.setterAnnotated = setterAnnotated
    }
```
That's not a one to one replacement like it should be.

[the remark this answers is not in the document]
Comment (vRallev) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt:259]: To be clear, that's my only concern for this PR. Everything else looks great!

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt (hunk 1 above)
Comment (ZacSweers) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/MembersInjectionUtils.kt:29]: ah, yes I believe we do that as-needed

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (ZacSweers) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/MembersInjectorGeneratorTest.kt]: 8f754c4

**Comment by vRallev** [comment]
Rebased the PR to fix the CI errors.

**Review by vRallev — APPROVED** [review]

**Comment by vRallev** [comment]
Damn, I screwed up. I meant to squash the commits... Life moves on.

**Pull request merged**